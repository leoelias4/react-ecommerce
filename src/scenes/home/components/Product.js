import React from "react";
import PropTypes from "prop-types";
import { css } from "react-emotion";

const styles = props => ({
  root: css`
    display: flex;
    flex-flow: column;
    position: relative;
    max-width: 345px;
    box-shadow: 0px 1px 5px 0px rgba(0, 0, 0, 0.2),
      0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12);
    overflow: hidden;
    padding: 10px;
    border-radius: 2px;
    background-color: #fff;
    transition: 0.3s all ease;
    label: product-root;
  `,
  image: css`
    background-color: #ababab1c;
    background-image: url(${props.picture});
    height: 0;
    padding-top: 100%;
    display: block;
    background-size: cover;
    background-repeat: no-repeat;
    background-position: center;
    label: image;
  `,
  details: css`
    display: grid;
    grid-template-rows: repeat(3, 50px);
    grid-gap: 5px;
    align-items: center;
    margin-top: 30px;
    align-items: center;
    justify-content: center;
    text-align: center;
    label: details;
  `,
  name: css`
    color: rgba(0, 0, 0, 0.87);
    font-size: 0.9rem;
    font-weight: 400;
    line-height: 1.35417em;
    margin: 0;
    display: block;
    label: name;
  `,
  itemsLeft: css`
    color: rgba(0, 0, 0, 0.87);
    font-size: 1rem;
    font-style: italic;
    color: #ababab;
    label: items-left;
  `,
  price: css`
    font-size: 2rem;
    color: #38c538;
    label: price;
  `,
  button: css`
    width: 56px;
    height: 56px;
    padding: 0;
    min-width: 0;
    font-size: 24px;
    box-shadow: 0px 3px 5px -1px rgba(0, 0, 0, 0.2),
      0px 6px 10px 0px rgba(0, 0, 0, 0.14), 0px 1px 18px 0px rgba(0, 0, 0, 0.12);
    border-radius: 50%;
    border: none;
    color: #fff;
    background-color: #2196f3;
    position: absolute;
    right: 10px;
    bottom: 10px;
    display: flex;
    justify-content: center;
    align-items: center;
    transition: 0.3s all ease;
    outline: none;
    cursor: pointer;
    label: button;
    &:hover {
      background-color: #1976d2;
    }
    &:disabled {
      background-color: #ababab;
      cursor: not-allowed;
    }
    &:active {
      box-shadow: none;
    }
  `,
  icon: css`
    fill: #fff;
    width: 1em;
    height: 1em;
    display: inline-block;
    font-size: 24px;
    transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    user-select: none;
    flex-shrink: 0;
    label: icon;
  `
});

const Product = props => {
  const style = styles(props);
  const s = props.quantity > 1 ? "s" : "";
  return (
    <div className={ style.root }>
      <div className={ style.image } />
      <div className={ style.details }>
        <span className={ style.name }>{ props.title }</span>
        <span className={ style.itemsLeft }>
          { props.quantity } item{ s } left
        </span>
        <span className={ style.price }>R${ props.price }</span>
      </div>
      <button className={ style.button } disabled={ props.quantity === 0 } onClick={ props.onClick }>
        <svg className={ style.icon } focusable="false" viewBox="0 0 24 24" aria-hidden="true">
          <g>
            <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />
          </g>
        </svg>
      </button>
    </div>
    );
};

export default Product;

Product.defaultProps = {
  title: "Product",
  quantity: 0,
  price: 0
};

Product.propTypes = {
  title: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  picture: PropTypes.string,
  price: PropTypes.number.isRequired,
  onClick: PropTypes.func
};
