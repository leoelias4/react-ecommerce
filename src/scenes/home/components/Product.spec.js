import React from "react";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Product from "./Product";
configure({ adapter: new Adapter() });

const setup = (title, quantity, price) => {
  const actions = {
    onClick: jest.fn()
  };

  const component = mount(
    <Product
      title={title}
      quantity={quantity}
      price={price}
      {...actions}
    />
  );
  return {
    component,
    actions,
    button: component.find("button").first()
  };
};

describe("Product component", () => {
  it("should render successfully", () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });

  it("should render cartitem title", () => {
    const { component } = setup("1");
    expect(component.props().title).toEqual("1");
  });
  
  it("should render cartitem without props", () => {
    const { component } = setup();
    expect(component.props().price).toEqual(0);
    expect(component.props().quantity).toEqual(0);
    expect(component.props().title).toEqual("Product");
  });

  it("should not call action on button click with 0 item left", () => {
    const { button, actions } = setup();
    button.simulate("click");
    expect(actions.onClick).not.toBeCalled();
  });

  it("should not call action on button click with more than 1 item left", () => {
    const { button, actions } = setup(undefined, 10);
    button.simulate("click");
    expect(actions.onClick).toBeCalled();
  });
});
