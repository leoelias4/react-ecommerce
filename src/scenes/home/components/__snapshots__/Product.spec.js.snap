// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component should render successfully 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Product
    onClick={[Function]}
    price={0}
    quantity={0}
    title="Product"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "onClick": [Function],
      "price": 0,
      "quantity": 0,
      "title": "Product",
    },
    "ref": null,
    "rendered": Object {
      "instance": <div
        class="css-16a2rgp-product-root"
>
        <div
                class="css-vqour5-image"
        />
        <div
                class="css-3s4mfs-details"
        >
                <span
                        class="css-wbrlco-name"
                >
                        Product
                </span>
                <span
                        class="css-1bk0x0y-items-left"
                >
                        0
                        item
                        left
                </span>
                <span
                        class="css-156zn4r-price"
                >
                        R$
                        0
                </span>
        </div>
        <button
                class="css-1wtfgww-button"
                disabled=""
        >
                <svg
                        class="css-12sbw6q-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                        aria-hidden="true"
                >
                        <g>
                                <path
                                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                />
                        </g>
                </svg>
        </button>
</div>,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="css-vqour5-image"
/>,
          <div
            className="css-3s4mfs-details"
>
            <span
                        className="css-wbrlco-name"
            >
                        Product
            </span>
            <span
                        className="css-1bk0x0y-items-left"
            >
                        0
                         item
                        
                         left
            </span>
            <span
                        className="css-156zn4r-price"
            >
                        R$
                        0
            </span>
</div>,
          <button
            className="css-1wtfgww-button"
            disabled={true}
            onClick={[Function]}
>
            <svg
                        aria-hidden="true"
                        className="css-12sbw6q-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
            >
                        <g>
                                    <path
                                                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                    />
                        </g>
            </svg>
</button>,
        ],
        "className": "css-16a2rgp-product-root",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": <div
            class="css-vqour5-image"
/>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "css-vqour5-image",
          },
          "ref": null,
          "rendered": Array [
            undefined,
          ],
          "type": "div",
        },
        Object {
          "instance": <div
            class="css-3s4mfs-details"
>
            <span
                        class="css-wbrlco-name"
            >
                        Product
            </span>
            <span
                        class="css-1bk0x0y-items-left"
            >
                        0
                        item
                        left
            </span>
            <span
                        class="css-156zn4r-price"
            >
                        R$
                        0
            </span>
</div>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="css-wbrlco-name"
>
                Product
</span>,
              <span
                className="css-1bk0x0y-items-left"
>
                0
                 item
                
                 left
</span>,
              <span
                className="css-156zn4r-price"
>
                R$
                0
</span>,
            ],
            "className": "css-3s4mfs-details",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": <span
                class="css-wbrlco-name"
>
                Product
</span>,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Product",
                "className": "css-wbrlco-name",
              },
              "ref": null,
              "rendered": Array [
                "Product",
              ],
              "type": "span",
            },
            Object {
              "instance": <span
                class="css-1bk0x0y-items-left"
>
                0
                item
                left
</span>,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  0,
                  " item",
                  "",
                  " left",
                ],
                "className": "css-1bk0x0y-items-left",
              },
              "ref": null,
              "rendered": Array [
                "0",
                " item",
                "",
                " left",
              ],
              "type": "span",
            },
            Object {
              "instance": <span
                class="css-156zn4r-price"
>
                R$
                0
</span>,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "R$",
                  0,
                ],
                "className": "css-156zn4r-price",
              },
              "ref": null,
              "rendered": Array [
                "R$",
                "0",
              ],
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": <button
            class="css-1wtfgww-button"
            disabled=""
>
            <svg
                        class="css-12sbw6q-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                        aria-hidden="true"
            >
                        <g>
                                    <path
                                                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                    />
                        </g>
            </svg>
</button>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <svg
              aria-hidden="true"
              className="css-12sbw6q-icon"
              focusable="false"
              viewBox="0 0 24 24"
>
              <g>
                            <path
                                          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                            />
              </g>
</svg>,
            "className": "css-1wtfgww-button",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": <svg
                class="css-12sbw6q-icon"
                focusable="false"
                viewBox="0 0 24 24"
                aria-hidden="true"
>
                <g>
                                <path
                                                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                />
                </g>
</svg>,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "aria-hidden": "true",
                "children": <g>
                  <path
                                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
</g>,
                "className": "css-12sbw6q-icon",
                "focusable": "false",
                "viewBox": "0 0 24 24",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": <g>
                    <path
                                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                    />
</g>,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <path
                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
/>,
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
/>,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "d": "M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z",
                      },
                      "ref": null,
                      "rendered": Array [
                        undefined,
                      ],
                      "type": "path",
                    },
                  ],
                  "type": "g",
                },
              ],
              "type": "svg",
            },
          ],
          "type": "button",
        },
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "onClick": [Function],
        "price": 0,
        "quantity": 0,
        "title": "Product",
      },
      "ref": null,
      "rendered": Object {
        "instance": <div
          class="css-16a2rgp-product-root"
>
          <div
                    class="css-vqour5-image"
          />
          <div
                    class="css-3s4mfs-details"
          >
                    <span
                              class="css-wbrlco-name"
                    >
                              Product
                    </span>
                    <span
                              class="css-1bk0x0y-items-left"
                    >
                              0
                              item
                              left
                    </span>
                    <span
                              class="css-156zn4r-price"
                    >
                              R$
                              0
                    </span>
          </div>
          <button
                    class="css-1wtfgww-button"
                    disabled=""
          >
                    <svg
                              class="css-12sbw6q-icon"
                              focusable="false"
                              viewBox="0 0 24 24"
                              aria-hidden="true"
                    >
                              <g>
                                        <path
                                                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                        />
                              </g>
                    </svg>
          </button>
</div>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="css-vqour5-image"
/>,
            <div
              className="css-3s4mfs-details"
>
              <span
                            className="css-wbrlco-name"
              >
                            Product
              </span>
              <span
                            className="css-1bk0x0y-items-left"
              >
                            0
                             item
                            
                             left
              </span>
              <span
                            className="css-156zn4r-price"
              >
                            R$
                            0
              </span>
</div>,
            <button
              className="css-1wtfgww-button"
              disabled={true}
              onClick={[Function]}
>
              <svg
                            aria-hidden="true"
                            className="css-12sbw6q-icon"
                            focusable="false"
                            viewBox="0 0 24 24"
              >
                            <g>
                                          <path
                                                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                          />
                            </g>
              </svg>
</button>,
          ],
          "className": "css-16a2rgp-product-root",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": <div
              class="css-vqour5-image"
/>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "css-vqour5-image",
            },
            "ref": null,
            "rendered": Array [
              undefined,
            ],
            "type": "div",
          },
          Object {
            "instance": <div
              class="css-3s4mfs-details"
>
              <span
                            class="css-wbrlco-name"
              >
                            Product
              </span>
              <span
                            class="css-1bk0x0y-items-left"
              >
                            0
                            item
                            left
              </span>
              <span
                            class="css-156zn4r-price"
              >
                            R$
                            0
              </span>
</div>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="css-wbrlco-name"
>
                  Product
</span>,
                <span
                  className="css-1bk0x0y-items-left"
>
                  0
                   item
                  
                   left
</span>,
                <span
                  className="css-156zn4r-price"
>
                  R$
                  0
</span>,
              ],
              "className": "css-3s4mfs-details",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": <span
                  class="css-wbrlco-name"
>
                  Product
</span>,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Product",
                  "className": "css-wbrlco-name",
                },
                "ref": null,
                "rendered": Array [
                  "Product",
                ],
                "type": "span",
              },
              Object {
                "instance": <span
                  class="css-1bk0x0y-items-left"
>
                  0
                  item
                  left
</span>,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    0,
                    " item",
                    "",
                    " left",
                  ],
                  "className": "css-1bk0x0y-items-left",
                },
                "ref": null,
                "rendered": Array [
                  "0",
                  " item",
                  "",
                  " left",
                ],
                "type": "span",
              },
              Object {
                "instance": <span
                  class="css-156zn4r-price"
>
                  R$
                  0
</span>,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "R$",
                    0,
                  ],
                  "className": "css-156zn4r-price",
                },
                "ref": null,
                "rendered": Array [
                  "R$",
                  "0",
                ],
                "type": "span",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": <button
              class="css-1wtfgww-button"
              disabled=""
>
              <svg
                            class="css-12sbw6q-icon"
                            focusable="false"
                            viewBox="0 0 24 24"
                            aria-hidden="true"
              >
                            <g>
                                          <path
                                                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                          />
                            </g>
              </svg>
</button>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <svg
                aria-hidden="true"
                className="css-12sbw6q-icon"
                focusable="false"
                viewBox="0 0 24 24"
>
                <g>
                                <path
                                                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                />
                </g>
</svg>,
              "className": "css-1wtfgww-button",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": <svg
                  class="css-12sbw6q-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                  aria-hidden="true"
>
                  <g>
                                    <path
                                                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                    />
                  </g>
</svg>,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "aria-hidden": "true",
                  "children": <g>
                    <path
                                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                    />
</g>,
                  "className": "css-12sbw6q-icon",
                  "focusable": "false",
                  "viewBox": "0 0 24 24",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": <g>
                      <path
                                            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
</g>,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
/>,
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": <path
                          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
/>,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "d": "M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z",
                        },
                        "ref": null,
                        "rendered": Array [
                          undefined,
                        ],
                        "type": "path",
                      },
                    ],
                    "type": "g",
                  },
                ],
                "type": "svg",
              },
            ],
            "type": "button",
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
